yalafi/packages/cleveref.py

```diff
@@ -0,0 +1,215 @@
+        return utils.latex_error('could not read file ' + repr(file), pos, parser.latex, parser.parms)
+
+    refs = {'\\cref': {'':{}, '*': {}}, '\\Cref': {'':{}, '*': {}}, '\\crefrange': {'':{}, '*': {}}, '\\Crefrange': {'':{}, '*': {}}}
+
+    for rep in sed.split('\n'):
+        # only consider non-empty lines:
+        if rep == '':
+            continue
+
+        # Match \cref,\cref*,\Cref and \Cref* and save the replacement string:
+        m = re_ref.match(rep)
+        if m:
+            refs[m.group(1)][m.group(2)][m.group(3)] = re_remove_escaped_symbols(m.group(4))
+            continue
+
+        # Match \crefrange, \crefrange*, \Crefrange and \Crefrange* and
+        # save the replacement string:
+        m = re_ref_range.match(rep)
+        if m:
+            refs[m.group(1)][m.group(2)][(m.group(3), m.group(4))] = re_remove_escaped_symbols(m.group(5))
+
+        # Match any other command and create Macro objects for them.
+        # See definition of re_command for more details:
+        m = re_command.match(rep)
+        if m:
+            args = 'A'*int((m.end(3)-m.start(3))/4)
+            string = re_remove_escaped_symbols(m.group(4))
+            if m.group(2):
+                name = re_cC.sub('c', m.group(1))
+                parser.the_macros[name] = Macro(parser.parms, name, args=args, repl=string)
+                name = re_cC.sub('C', m.group(1))
+                parser.the_macros[name] = Macro(parser.parms, name, args=args, repl=string)
+            else:
+                name = m.group(1)
+                parser.the_macros[name] = Macro(parser.parms, name, args=args, repl=string)
+
+    # Make the \cref, …, \Crefrange* Macro objects:
+    for ref in ['\\cref','\\Cref']:
+        parser.the_macros[ref] = Macro(parser.parms, ref, args='*A', repl=h_make_cref(refs[ref]))
+    for ref in ['\\crefrange','\\Crefrange']:
+        parser.the_macros[ref] = Macro(parser.parms, ref, args='*AA', repl=h_make_crefrange(refs[ref]))
+
+    # \YYCleverefInput should not produce any output:
+    return []
+
+
+def h_make_cref(cref):
+    def f(parser, buf, mac, args, delim, pos):
+        star = parser.get_text_direct(args[0])
+        rep = parser.get_text_direct(args[1])
+        if rep in cref[star]:
+            toks = parser.parms.scanner.scan(cref[star][rep])
+            for t in toks:
+                t.pos = pos
+            return toks
+        return utils.latex_error(f'Reference {rep} for command {mac.name} undefined.\n*** Run LaTeX again to build a new .sed file', pos, parser.latex, parser.parms)
+    return f
+
+
+def h_make_crefrange(cref):
+    def f(parser, buf, mac, args, delim, pos):
+        star = parser.get_text_direct(args[0])
+        rep = (parser.get_text_direct(args[1]), parser.get_text_direct(args[2]))
+        if rep in cref[star]:
+            toks = parser.parms.scanner.scan(cref[star][rep])
+            for t in toks:
+                t.pos = pos
+            return toks
+        return []

```

Comment:
I really like the way you catch all possible usage mistakes. Could you raise a message here as in `h_make_sref()`? (Just for the case that the `cleveref` issue is fixed, sometime.)